subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -889,20 +858,281 @@ static int nrf_cloud_encode_service_info_ui(const struct nrf_cloud_svc_info_ui *
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int encode_modem_info_sim(struct sim_param *sim, cJSON *json_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(sim != NULL);
+	__ASSERT_NO_MSG(json_obj != NULL);
+
+	ret = add_modem_info_data(&sim->uicc, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&sim->iccid, json_obj);
+	if (ret) {
+		LOG_DBG("sim_param object does not contain an ICCID")
+	}
+
+	ret = add_modem_info_data(&sim->imsi, json_obj);
+	if (ret) {
+		LOG_DBG("sim_param object does not contain an IMSI")
+	}
+
+	return 0;
+}
+
+static int encode_modem_info_device(struct device_param *device, cJSON *json_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(device != NULL);
+	__ASSERT_NO_MSG(json_obj != NULL);
+
+	ret = add_modem_info_data(&device->modem_fw, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&device->battery, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&device->imei, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "board", device->board) == NULL) {
+		return -EINVAL;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "appVersion", device->app_version) == NULL) {
+		return -EINVAL;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "appName", device->app_name) == NULL) {
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int encode_modem_info_json_object(struct modem_param_info *modem,
+				  cJSON *root_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(root_obj != NULL);
+	__ASSERT_NO_MSG(modem != NULL);
+
+	int obj_count = cJSON_GetArraySize(root_obj);
+
+	cJSON *network_obj = NULL;
+	cJSON *sim_obj = NULL;
+	cJSON *device_obj = NULL;
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_NETWORK)) {
+		network_obj = cJSON_CreateObject();
+		if (network_obj == NULL) {
+			return -ENOMEM;
+		}
+		ret = encode_modem_info_network(&modem->network, network_obj);
+		if (!ret) {
+			ret = json_add_obj_cs(root_obj, NRF_CLOUD_DEVICE_JSON_KEY_NET_INF, network_obj);
+			if (ret) {
+				cJSON_Delete(network_obj);
+				return -ENOMEM;
+			}
+		} else {
+			cJSON_Delete(network_obj);
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_SIM)) {
+		sim_obj = cJSON_CreateObject();
+		if (sim_obj == NULL) {
+			return -ENOMEM;
+		}
+		ret = encode_modem_info_sim(&modem->sim, sim_obj);
+		if (!ret) {
+			ret = json_add_obj_cs(root_obj, NRF_CLOUD_DEVICE_JSON_KEY_SIM_INF, sim_obj);
+			if (ret) {
+				cJSON_Delete(sim_obj);
+				return -ENOMEM;
+			}
+		} else {
+			cJSON_Delete(sim_obj);
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_DEVICE)) {
+		device_obj = cJSON_CreateObject();
+		if (device_obj == NULL) {
+			return -ENOMEM;
+		}
+		ret = encode_modem_info_device(&modem->device, device_obj);
+		if (!ret) {
+			ret = json_add_obj_cs(root_obj, NRF_CLOUD_DEVICE_JSON_KEY_DEV_INF, device_obj);
+			if (ret) {
+				cJSON_Delete(device_obj);
+				return -ENOMEM;
+			}
+		} else {
+			cJSON_Delete(device_obj);
+		}
+	}
+
+	return obj_count = cJSON_GetArraySize(root_obj) - obj_count;
```

Comment:
option: remove `obj_count` and just have the function return zero on success and negative on error.

Reply:
done